torch/csrc/distributed/rpc/tensorpipe_utils.h

```diff
@@ -97,15 +99,32 @@ struct TORCH_CUDA_CPP_API CudaLazyStreamContext : public LazyStreamContext {
     if (iter == streams_.end()) {
       auto cudaStream = at::cuda::getStreamFromPool(
           /* isHighPriority */ false, /* device */ index);
+      for (auto& barrierEvent : barrierEvents_) {
+        barrierEvent.block(cudaStream);
```

Comment:
I am not sure if I understand this part.  When we reach here, there has been no stream_ for given device  in streams_ array. So all barierEvent are for other gpus. How you can block a steam of one device on events in streams of other devices?

Reply:
> How you can block a steam of one device on events in streams of other devices?

IIUC, events on one device can block streams from a different device. Quote from CUDAEvent doc:

> The underlying CUDA events are lazily initialized when the event is first recorded or exported to another process. After creation, only streams on the same device may record the event. However, streams on any device can wait on the event.

Reply:
I think the way this code ended up being written is very "implicit" and thus hard to understand at first glance. If I got it right, this method is called at two separate stages on the same object (first during allocation of the incoming request, then during serialization of the outgoing response), and in these two occurrences the circumstances are quite different: the first time the barrier events are empty, hence this extra code has no effect, whereas the second time there are barrier events, but this code is only executed for streams that were not already accessed the first time.

Hence only _new_ devices that were added by the callback end up waiting for barrier events, thus overall I think it may work, but it's rather hard to follow as the same code behaves very differently in different cases, based on "distant interactions".

Reply:
> I think the way this code ended up being written is very "implicit" and thus hard to understand at first glance. 

Yep, this is a temporary solution to remove flakiness from master. I would like kicking off these tests again and see if this PR fix the flakiness. I am not confident this is the only problem we have. 

As we discussed offline, I was thinking about a long solution as below:

```python
@rpc.functions.enable_cuda(devices=[0, 1])
def user_func(x):
  assert(x.device = "cuda:0")
  return x.to(1)
```

This API should be sufficient to allow the remote side to synchronize streams correctly. And when this is in, I will remove the current solution in this PR. I didn't go that way initially because I don't wanna introduce API changes before we reach consensus on that. I will the similar solution could also work for `CUDAFuture`?

> I also want to reiterate my thinking that if users fully embrace RPC they should never need to manually move tensors to other devices, as any such move could be replaced by a RPC call, thus meaning that each remote RPC function just takes care of processing on a fixed set of devices.

Yep, we can also do this initially. If we all agree this is better solution for now, I will disable all cross-device tests.

Reply:
We can also be explicit on this. If the function returns a tensor on a different device but the function wasn't decorated that way, we just error out.